packages/databus/src/index.ts

```diff
@@ -0,0 +1,133 @@
+/*-----------------------------------------------------------------------------
+| Copyright (c) Jupyter Development Team.
+| Distributed under the terms of the Modified BSD License.
+|----------------------------------------------------------------------------*/
+
+import { DisposableDelegate, IDisposable } from '@phosphor/disposable';
+
+import { ArrayExt } from '@phosphor/algorithm';
+
+import { ISignal, Signal } from '@phosphor/signaling';
+
+/**
+ * An interface for an abstract dataset, with a mimetype, data, and metadata.
+ *
+ * #### Notes
+ * This interface is similar conceptually to that of nbformat.IMimeBundle,
+ * but with only a single mimetype and different structure. We expect there
+ * to be utilities to convert between this formats.
+ */
+export interface IDataset {
+  /**
+   * The string mimetype for the dataset.
+   */
+  mimeType: string;
+
+  /***
+   * The data of the dataset.
+   *
+   * #### Notes
+   * The notion of data in the databus is completely abstract, and typed as an
+   * any. Publishers and consumers of a given mimetype will need to agree
+   * what the actual data is and cast it appropriately before use.
+   */
+  data: any;
+
+  /**
+   * The metadata associate with the dataset.
+   *
+   * #### Notes
+   * Like the data itself, the databus notion of metadata is entirely abstract.
+   * We are currently exploring different metadata schemes, so this may change
+   * in the future.
+   */
+  metadata: any;
```

Comment:
Until we have some sort of typing for metadata, it might clearer not to include it or type it as an empty object. Otherwise, as a user, I am not sure to tell whether to put some part of the data in the metadata or in the actual data.

Reply:
Yeah, until we figure out more about the metadata, we may need to hedge a bit.

Reply:
Reading your comments on the issue, it seems like one way to proceed would be to say "A certain mimetype corresponds to a certain structure of data/metadata".

If we were thinking about this problem in Typescript land, that sounds a lot like a subclass. So could have a `Data` abstract base class, that requires mimetype, metadata and data properties, but metadata and data are generically typed.  Then you would subclass from that for CSV data, to narrow down possibilities of what is in those fields. [Here is an example](http://www.typescriptlang.org/play/#src=interface%20IData%3CT%2C%20V%3E%20%7B%0D%0A%20%20%20%20mimeType%3A%20String%0D%0A%20%20%20%20data%3A%20T%0D%0A%20%20%20%20metadata%3A%20V%0D%0A%7D%0D%0A%0D%0A%0D%0Ainterface%20ICSVData%3CT%3E%20extends%20IData%3CArray%3CT%3E%2C%20%7B%20%5BP%20in%20keyof%20T%5D%3A%20string%20%7D%3E%20%7B%0D%0A%20%20%20%20mimeType%3A%20%22CSV%22%2C%0D%0A%20%20%20%20data%3A%20Array%3CT%3E%2C%0D%0A%20%20%20%20metadata%3A%20%7B%20%5BP%20in%20keyof%20T%5D%3A%20string%20%7D%0D%0A%7D%0D%0A%0D%0Aconst%20c%3A%20ICSVData%20%3C%20%7B%20name%3A%20string%20%7D%20%3E%20%3D%20%20%7B%0D%0A%20%20%20%20mimeType%3A%20%22CSV%22%2C%0D%0A%20%20%20%20data%3A%20%5B%7B%20name%3A%20%22hi%22%20%7D%5D%2C%0D%0A%20%20%20%20metadata%3A%20%7B%22name%22%3A%20%22string%22%7D%0D%0A%7D%0D%0A%0D%0A%0D%0A%2F%2F%2F%20or%20could%20do%20class%20based%20approach%0D%0A%0D%0Aclass%20CSVData%3CT%3E%20implements%20IData%3CArray%3CT%3E%2C%20%7B%20%5BP%20in%20keyof%20T%5D%3A%20string%20%7D%3E%20%7B%0D%0A%20%20%20%20mimeType%20%3D%20%22CSV%22%3B%0D%0A%0D%0A%20%20%20%20constructor(public%20data%3A%20Array%3CT%3E%2C%20public%20metadata%3A%20%7B%20%5BP%20in%20keyof%20T%5D%3A%20string%20%7D)%20%7B%0D%0A%20%20%20%20%20%20%20%20%0D%0A%20%20%20%20%7D%0D%0A%7D%0D%0A%0D%0A%0D%0Aconst%20c2%3A%20CSVData%3C%7B%20name%3A%20string%20%7D%3E%20%3D%20new%20CSVData(%5B%7B%20name%3A%20%22hi%22%20%7D%5D%2C%20%7B%20%22name%22%3A%20%22string%22%20%7D)%3B)